### PR TITLE
Add DLR native support - part 1 

### PR DESCRIPTION
**Pull request by stu1130**
## Channel for questions ##
Add DLR native jni

## Description ##
(Brief description of what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- [ ] Code is well-documented: 
    - For user-facing API changes, Java doc has been updated. 
    - For new examples, README.md is added to explain the what the example does.
- [ ] To the my best knowledge, [examples](https://github.com/awslabs/djl/tree/master/examples) and [jupyter notebooks](https://github.com/awslabs/djl/tree/master/jupyter) are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, Java doc)
- [ ] Feature2, tests, (and when applicable, Java doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made?
- Interesting edge cases to note here

### CODE REVIEW

**Comment by lanking520** [comment]
Since DLR built binary is tiny, we can place the native along with the engine like SentencePiece. So we only publish the engine package.

File: dlr/dlr-engine/src/main/java/ai/djl/dlr/jni/NativeResource.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        }
+    }
+
+    /**
+     * Gets the boolean that indicates whether this resource has been released.
+     *
+     * @return whether this resource has been released
+     */
+    public boolean isReleased() {
+        return handle == null;
```
Comment (frankfliu) [inline dlr/dlr-engine/src/main/java/ai/djl/dlr/jni/NativeResource.java:43]: handler should never be null
```suggestion
        return handle.get() == null;
```

File: dlr/dlr-engine/src/main/java/ai/djl/dlr/jni/NativeResource.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+/**
+ * {@code NativeResource} is an internal class for {@link AutoCloseable} blocks of memory created in
+ * the DLR Engine.
+ */
+public abstract class NativeResource implements AutoCloseable {
+
+    private static final Logger logger = LoggerFactory.getLogger(NativeResource.class);
+
+    protected AtomicLong handle;
```
Comment (frankfliu) [inline dlr/dlr-engine/src/main/java/ai/djl/dlr/jni/NativeResource.java:27]:
```suggestion
    protected AtomicReference<Long> handle;
```

File: dlr/dlr-native/build.gradle [hunk 3]
```diff
@@ -0,0 +1,18 @@
+plugins {
+    id 'maven-publish'
+    id 'signing'
+}
+
+apply from: file("${rootProject.projectDir}/tools/gradle/cpp-formatter.gradle")
+
+task compileJNI {
+    doFirst {
+        if (System.properties['os.name'].toLowerCase(Locale.ROOT).contains("mac")) {
```
Comment (frankfliu) [inline dlr/dlr-native/build.gradle:10]: Only for Mac, not linux?

File: dlr/dlr-native/src/ai_djl_dlr_jni_DlrLibrary.cc [hunk 4]
```diff
@@ -0,0 +1,175 @@
+/*
```
Comment (frankfliu) [inline dlr/dlr-native/src/ai_djl_dlr_jni_DlrLibrary.cc:1]: should be in src/main/native/ folder

**Review by frankfliu — CHANGES_REQUESTED** [review]

**Comment by stu1130** [comment]
@lanking520 talked to @frankfliu offline we still want to separate the native jar and engine jar consider we would have lots of platforms to publish.

File: dlr/dlr-native/build.gradle (hunk 3 above)
Comment (stu1130) [reply dlr/dlr-native/build.gradle:10]: I will first make Mac work then extend to linux and other platform

**Review by lanking520 — APPROVED** [review]

**Review by frankfliu — APPROVED** [review]

**Pull request merged**